app/models/app_version.rb

```diff
@@ -0,0 +1,48 @@
+class AppVersion
+  include Comparable
+
+  DECIMAL_INTEGER = /(?:0|[1-9]\d*)/
+  VERSION_REGEX = /\A(?:#{DECIMAL_INTEGER}\.){0,2}#{DECIMAL_INTEGER}\z/ # iOS rules currently, but can be relaxed from here
```

Comment:
So iOS seems to be explicitly 3 integers always (at least from [reading their docs](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-111349)). I could be wrong on that.

Seems like android is the one that might need it relaxed. [Their docs](https://developer.android.com/studio/publish/versioning) make it seem like `versionName` is _intended_ to follow (major).(minor).(patch) formatting, but then their example doesn't.

Anyway, what's the story for Android support here? Especially if we leave the land of well-formed integers? Can we still force funkier version formats into a pg array comparison?

Reply:
See the tests - Android will proffer the monotonic integer build thingie instead of a version number.

Reply:
Also agreed - this supports up to 3 integers like the Apple spec because the Apple spec says so. :) am I missing something?

Reply:
I guess my question was more how Android enforces constraints on top-level `versionName` values. Like, is there something stopping me from releasing versions like `1.2-beta1` and `1.2-beta6`?

Reply:
(In their example they have a `1.1-demo` as one of the "flavors" of release, wasn't sure if the formatting gets stricter outside of the flavors.)

Reply:
We won’t touch that number, Android has a second secret build number integer they require each build to sport so they don’t even bother to have collatable version numbers for sorting in the store. we’ll ingest the integer.

Reply:
I’d only be interested in adding text support for a platform like a desktop app or something that had a richer convention around prerelease versions but still had a formal collation algorithm.

Reply:
Also, Here's the apple technical note I referenced - they are apparently down with one-to-three integers, not just exactly three, so for the time being, I think I have algo parity with Apple (though I'm open to relaxing it further to support other platforms - it just so happened that Android `versionCode`s are a subset of the Apple spec): https://developer.apple.com/library/archive/technotes/tn2420/_index.html#//apple_ref/doc/uid/DTS40016603-CH1-NUMBERING_CONVENTIONS-EXAMPLES

I'm going to link this tech note into the comments of the REGEX too.